File: BuySell2.java
// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/*
    We basically buy all all the dips and hold it until the stock price keeps rising.
*/

class Solution {
    public int maxProfit(int[] prices) {
        
        int profit = 0;
        for(int i=1; i<prices.length; i++){
            if (prices[i-1]<prices[i])
                profit += prices[i]-prices[i-1];
            
        }
        return profit;
    }
}

File: Buy_sell_2.py
# Time complexity : O(n)
# Space complexityy : O(1)
# Leetcode : Solved and submitted

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices) - 1
        profit = 0
        valley = peak = prices[0]
        i = 0
        
        # traverse over the array
        while i < n:
           # find the valley which price doing down, which is the least price
            while i < n and prices[i] >= prices[i+1]:
                i += 1
            valley = prices[i]
            
            # find the highest price going ahead
            while i < n and prices[i] <= prices[i+1]:
                i += 1
            peak = prices[i]
            
            # fetch the profit
            profit += peak - valley
        
        return profit
